File: solutions/day03/solution.clj
(def steps (seq (slurp "input.txt")))

(defn move
  [pos step]
  (let [coord (if (#{\>, \<} step) :x :y)
        step (if (#{\>, \^} step) 1 -1)]
    (assoc pos coord (+ (coord pos) step))))

(defn route
  [steps]
  (loop [[step & more] steps
         pos {:x 0 :y 0}
         route []]
    (let [route (conj route pos)]
      (if step
        (recur more 
               (move pos step) 
               (conj route pos))
        route))))

(defn num-distinct
  [route]
  (count (set route)))

(defn part1
  [steps]
  (num-distinct (route steps)))

(defn part2
  [steps]
  (let [santa (route (take-nth 2 steps))
        robo-santa (route (take-nth 2 (rest steps)))]
    (num-distinct (concat santa robo-santa))))

(println "part 1:" (part1 steps))
(println "part 2:" (part2 steps))

File: solutions/day03/solution.scala
val path = scala.io.Source.fromFile("input.txt").toList

val coord = Map('>' -> 'x',
                '<' -> 'x',
                '^' -> 'y',
                'v' -> 'y')

val dir = Map('>' -> 1,
              '<' -> -1,
              '^' -> 1,
              'v' -> -1)

def locationsVisited(steps: List[Char]) = {
    val pos = collection.mutable.Map('x' -> 0,
                                     'y' -> 0)
    val visited = collection.mutable.Set[Any]()
    for (step <- steps) {
        pos(coord(step)) += dir(step) 
        visited += pos.clone()
    }
    visited
}

val visited = locationsVisited(path)
println(s"part 1: ${visited.size}")

val grouped = path.grouped(2).toList
val santaSteps = grouped.map(_(0))
val robotSteps = grouped.map(_(1))
val allVisited = locationsVisited(santaSteps) | locationsVisited(robotSteps)
println(s"part 2: ${allVisited.size}")
